README.md
# embedded-sfa3x

Embedded I2C and UART drivers for the Sensirion sfa3x sensor.
The I2C driver can be found in the `i2c` directory and the UART driver in
the `uart` directory.

These drivers are not meant to be used at the same time.

To compile the example usage of the drivers follow the instructions in the
README file in the corresponding folder:
 * [./i2c/README.md](i2c/README.md)
 * [./uart/README.md](uart/README.md)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Fixed

 * uart: Handle SHDLC execution errors
 * i2c: Fixed SW / bit banging I2C sample implementation

### Removed

 * Remove Arduino sample implementations since we provide Arduino drivers

## [0.1.1] - 2021-08-01

### Changed

 * Increase read measurement delay to 5ms

## [0.1.0] - 2020-12-11

Initial release

[0.1.1]: https://github.com/Sensirion/embedded-sfa3x/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/Sensirion/embedded-sfa3x/releases/tag/0.1.0


i2c/tests/sfa3x-test.c
/*
 * Copyright (c) 2020, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sensirion_common.h"
#include "sensirion_i2c.h"
#include "sensirion_i2c_hal.h"
#include "sensirion_test_setup.h"
#include "sfa3x_i2c.h"
#include <inttypes.h>
#include <stdio.h>

TEST_GROUP (sfa3x_Tests) {
    void setup() {
        sensirion_i2c_hal_init();
    }

    void teardown() {
        int16_t error;
        error = sfa3x_reset_device();
        CHECK_ZERO_TEXT(error, "sfa3x_device_reset");
        sensirion_i2c_hal_free();
    }
};

TEST (sfa3x_Tests, sfa3x_Test_start_continuous_measurement) {
    int16_t error;
    error = sfa3x_start_continuous_measurement();
    CHECK_ZERO_TEXT(error, "sfa3x_start_continuous_measurement");
}

TEST (sfa3x_Tests, sfa3x_Test_stop_measurement) {
    int16_t error;
    error = sfa3x_stop_measurement();
    CHECK_ZERO_TEXT(error, "sfa3x_stop_measurement");
}

TEST (sfa3x_Tests, sfa3x_Test_read_measured_values) {
    int16_t error;
    int16_t hcho;
    int16_t humidity;
    int16_t temperature;
    error = sfa3x_read_measured_values(&hcho, &humidity, &temperature);
    CHECK_ZERO_TEXT(error, "sfa3x_read_measured_values");
    printf("hcho: %i\n", hcho);
    printf("humidity: %i\n", humidity);
    printf("temperature: %i\n", temperature);
}

TEST (sfa3x_Tests, sfa3x_Test_get_device_marking) {
    int16_t error;
    uint8_t device_marking[42];
    uint8_t device_marking_size = 42;
    error = sfa3x_get_device_marking(&device_marking[0], device_marking_size);
    CHECK_ZERO_TEXT(error, "sfa3x_get_device_marking");
    printf("device_marking: %s\n", device_marking);
}

TEST (sfa3x_Tests, sfa3x_Test_device_reset) {
    int16_t error;
    error = sfa3x_device_reset();
    CHECK_ZERO_TEXT(error, "sfa3x_device_reset");
}


uart/tests/sfa3x-test.c
/*
 * Copyright (c) 2020, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sensirion_common.h"
#include "sensirion_shdlc.h"
#include "sensirion_test_setup.h"
#include "sensirion_uart_hal.h"
#include "sfa3x_uart.h"
#include <inttypes.h>
#include <stdio.h>

TEST_GROUP (sfa3x_Tests) {
    void setup() {
        sensirion_uart_hal_init();
    }

    void teardown() {
        int16_t error;
        error = sfa3x_reset_device();
        CHECK_ZERO_TEXT(error, "sfa3x_device_reset");
        sensirion_uart_hal_free();
    }
};

TEST (sfa3x_Tests, sfa3x_Test_start_continuous_measurement) {
    int16_t error;
    error = sfa3x_start_continuous_measurement();
    CHECK_ZERO_TEXT(error, "sfa3x_start_continuous_measurement");
}

TEST (sfa3x_Tests, sfa3x_Test_stop_measurement) {
    int16_t error;
    error = sfa3x_stop_measurement();
    CHECK_ZERO_TEXT(error, "sfa3x_stop_measurement");
}

TEST (sfa3x_Tests, sfa3x_Test_read_measured_values_output_format_2) {
    int16_t error;
    int16_t hcho;
    int16_t relative_humidity;
    int16_t temperature;
    error = sfa3x_read_measured_values_output_format_2(
        &hcho, &relative_humidity, &temperature);
    CHECK_ZERO_TEXT(error, "sfa3x_read_measured_values_output_format_2");
    printf("hcho: %i\n", hcho);
    printf("relative_humidity: %i\n", relative_humidity);
    printf("temperature: %i\n", temperature);
}

TEST (sfa3x_Tests, sfa3x_Test_device_reset) {
    int16_t error;
    error = sfa3x_device_reset();
    CHECK_ZERO_TEXT(error, "sfa3x_device_reset");
}


i2c/tests/sfa3x_i2c_test.c
/**
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * Generator:           0.2.0
 * Yaml Version:        0.1.0
 * Template Version:    0.1.1
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sensirion_common.h"
#include "sensirion_i2c.h"
#include "sensirion_i2c_hal.h"
#include "sensirion_test_setup.h"
#include "sfa3x_i2c.h"
#include <inttypes.h>
#include <stdio.h>

TEST_GROUP (SFA3X_Tests) {
    void setup() {
        sensirion_i2c_hal_init();
    }

    void teardown() {

        int16_t error;
        error = sfa3x_device_reset();
        CHECK_ZERO_TEXT(error, "sfa3x_device_reset");

        sensirion_i2c_hal_free();
    }
};

TEST (SFA3X_Tests, SFA3X_Test_start_continuous_measurement) {
    int16_t error;
    error = sfa3x_start_continuous_measurement();
    CHECK_ZERO_TEXT(error, "sfa3x_start_continuous_measurement");
}

TEST (SFA3X_Tests, SFA3X_Test_stop_measurement) {
    int16_t error;
    error = sfa3x_stop_measurement();
    CHECK_ZERO_TEXT(error, "sfa3x_stop_measurement");
}

TEST (SFA3X_Tests, SFA3X_Test_read_measured_values) {
    int16_t error;
    int16_t hcho;
    int16_t humidity;
    int16_t temperature;
    error = sfa3x_read_measured_values(&hcho, &humidity, &temperature);
    CHECK_ZERO_TEXT(error, "sfa3x_read_measured_values");
    printf("hcho: %i\n", hcho);
    printf("humidity: %i\n", humidity);
    printf("temperature: %i\n", temperature);
}

TEST (SFA3X_Tests, SFA3X_Test_get_device_marking) {
    int16_t error;
    uint8_t device_marking[42];
    uint8_t device_marking_size = 42;
    error = sfa3x_get_device_marking(&device_marking[0], device_marking_size);
    CHECK_ZERO_TEXT(error, "sfa3x_get_device_marking");
    printf("device_marking: %s\n", device_marking);
}

TEST (SFA3X_Tests, SFA3X_Test_device_reset) {
    int16_t error;
    error = sfa3x_device_reset();
    CHECK_ZERO_TEXT(error, "sfa3x_device_reset");
}


i2c/sfa3x-i2c.c
/*
 * Copyright (c) 2020, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sfa3x-i2c.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"
#include "sensirion_i2c_hal.h"

int16_t sfa3x_start_continuous_measurement() {
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x06);

    return sensirion_i2c_write_data(93, &buffer[0], offset);
}

int16_t sfa3x_stop_measurement() {
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x104);

    return sensirion_i2c_write_data(93, &buffer[0], offset);
}

int16_t sfa3x_read_measured_values(int16_t* hcho, int16_t* humidity,
                                   int16_t* temperature) {
    int16_t error;
    uint8_t buffer[9];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x327);

    error = sensirion_i2c_write_data(93, &buffer[0], offset);
    if (error) {
        return error;
    }

    sensirion_i2c_hal_sleep_usec(2000);

    error = sensirion_i2c_read_data_inplace(93, &buffer[0], 6);
    if (error) {
        return error;
    }
    *hcho = sensirion_common_bytes_to_int16_t(&buffer[0]);
    *humidity = sensirion_common_bytes_to_int16_t(&buffer[2]);
    *temperature = sensirion_common_bytes_to_int16_t(&buffer[4]);
    return NO_ERROR;
}

int16_t sfa3x_get_device_marking(uint8_t* device_marking,
                                 uint8_t device_marking_size) {
    int16_t error;
    uint8_t buffer[48];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0xD060);

    error = sensirion_i2c_write_data(93, &buffer[0], offset);
    if (error) {
        return error;
    }

    sensirion_i2c_hal_sleep_usec(2000);

    error = sensirion_i2c_read_data_inplace(93, &buffer[0], 32);
    if (error) {
        return error;
    }
    sensirion_common_copy_bytes(&buffer[0], device_marking,
                                device_marking_size);
    return NO_ERROR;
}

int16_t sfa3x_device_reset() {
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0xD304);

    return sensirion_i2c_write_data(93, &buffer[0], offset);
}


i2c/sfa3x_i2c.c
/**
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * Generator:           0.2.0
 * Yaml Version:        0.1.0
 * Template Version:    0.1.1-32-g825e6f9
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sfa3x_i2c.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"
#include "sensirion_i2c_hal.h"

int16_t sfa3x_start_continuous_measurement() {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x06);

    error = sensirion_i2c_write_data(93, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(1000);
    return NO_ERROR;
}

int16_t sfa3x_stop_measurement() {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x104);

    error = sensirion_i2c_write_data(93, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(50000);
    return NO_ERROR;
}

int16_t sfa3x_read_measured_values(int16_t* hcho, int16_t* humidity,
                                   int16_t* temperature) {
    int16_t error;
    uint8_t buffer[9];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x327);

    error = sensirion_i2c_write_data(93, &buffer[0], offset);
    if (error) {
        return error;
    }

    sensirion_i2c_hal_sleep_usec(5000);

    error = sensirion_i2c_read_data_inplace(93, &buffer[0], 6);
    if (error) {
        return error;
    }
    *hcho = sensirion_common_bytes_to_int16_t(&buffer[0]);
    *humidity = sensirion_common_bytes_to_int16_t(&buffer[2]);
    *temperature = sensirion_common_bytes_to_int16_t(&buffer[4]);
    return NO_ERROR;
}

int16_t sfa3x_get_device_marking(uint8_t* device_marking,
                                 uint8_t device_marking_size) {
    int16_t error;
    uint8_t buffer[48];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0xD060);

    error = sensirion_i2c_write_data(93, &buffer[0], offset);
    if (error) {
        return error;
    }

    sensirion_i2c_hal_sleep_usec(2000);

    error = sensirion_i2c_read_data_inplace(93, &buffer[0], 32);
    if (error) {
        return error;
    }
    sensirion_common_copy_bytes(&buffer[0], device_marking,
                                device_marking_size);
    return NO_ERROR;
}

int16_t sfa3x_device_reset() {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0xD304);

    error = sensirion_i2c_write_data(93, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(100000);
    return NO_ERROR;
}


i2c/sfa3x_i2c.h
/**
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * Generator:           0.2.0
 * Yaml Version:        0.1.0
 * Template Version:    0.1.1-32-g825e6f9
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SFA3X_I2C_H
#define SFA3X_I2C_H

#ifdef __cplusplus
extern "C" {
#endif

#include "sensirion_config.h"

/**
 * sfa3x_start_continuous_measurement() - Starts continuous measurement in
 * polling mode.
 *
 * @note This command is only available in idle mode.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sfa3x_start_continuous_measurement(void);

/**
 * sfa3x_stop_measurement() - Stops the measurement mode and returns to idle
 * mode.
 *
 * @note This command is only available in measurement mode.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sfa3x_stop_measurement(void);

/**
 * sfa3x_read_measured_values() - Returns the new measurement results as
 * integers.
 *
 * @param hcho Formaldehyde concentration in ppb with a scaling of 5.
 *
 * @param humidity Relative humidity in % RH with a scaling of 100.
 *
 * @param temperature Temperature in degrees Celsius with a scaling of 200.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sfa3x_read_measured_values(int16_t* hcho, int16_t* humidity,
                                   int16_t* temperature);

/**
 * sfa3x_get_device_marking() - Read the device marking string from the device.
 *
 * @param device_marking ASCII string containing the serial number. The string
 * has the null-termination character included.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sfa3x_get_device_marking(uint8_t* device_marking,
                                 uint8_t device_marking_size);

/**
 * sfa3x_device_reset() - Executes a reset on the device.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sfa3x_device_reset(void);

#ifdef __cplusplus
}
#endif

#endif /* SFA3X_I2C_H */


uart/sfa3x_uart.c
/**
 * THIS FILE IS AUTOMATICALLY GENERATED AND MUST NOT BE EDITED MANUALLY!
 *
 * Generator:           0.8.2
 * Yaml Version:        0.1.0
 * Template Version:    0.1.1-32-g825e6f9
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sfa3x_uart.h"
#include "sensirion_common.h"
#include "sensirion_shdlc.h"
#include "sensirion_uart_hal.h"

int16_t sfa3x_start_continuous_measurement() {
    struct sensirion_shdlc_rx_header header;
    uint8_t buffer[12];
    struct sensirion_shdlc_buffer frame;
    sensirion_shdlc_begin_frame(&frame, &buffer[0], 0x00, 0x00, 1);
    sensirion_shdlc_add_uint8_t_to_frame(&frame, 0x00);

    sensirion_shdlc_finish_frame(&frame);
    int16_t error = sensirion_shdlc_tx_frame(&frame);
    if (error) {
        return error;
    }

    sensirion_uart_hal_sleep_usec(10000);

    error = sensirion_shdlc_rx_inplace(&frame, 0, &header);
    if (error) {
        return error;
    }
    return NO_ERROR;
}

int16_t sfa3x_stop_measurement() {
    struct sensirion_shdlc_rx_header header;
    uint8_t buffer[12];
    struct sensirion_shdlc_buffer frame;
    sensirion_shdlc_begin_frame(&frame, &buffer[0], 0x01, 0x00, 0);

    sensirion_shdlc_finish_frame(&frame);
    int16_t error = sensirion_shdlc_tx_frame(&frame);
    if (error) {
        return error;
    }

    sensirion_uart_hal_sleep_usec(10000);

    error = sensirion_shdlc_rx_inplace(&frame, 0, &header);
    if (error) {
        return error;
    }
    return NO_ERROR;
}

int16_t sfa3x_read_measured_values_output_format_2(int16_t* hcho,
                                                   int16_t* relative_humidity,
                                                   int16_t* temperature) {
    struct sensirion_shdlc_rx_header header;
    uint8_t buffer[24];
    struct sensirion_shdlc_buffer frame;
    sensirion_shdlc_begin_frame(&frame, &buffer[0], 0x03, 0x00, 1);
    sensirion_shdlc_add_uint8_t_to_frame(&frame, 0x02);

    sensirion_shdlc_finish_frame(&frame);
    int16_t error = sensirion_shdlc_tx_frame(&frame);
    if (error) {
        return error;
    }

    sensirion_uart_hal_sleep_usec(100000);

    error = sensirion_shdlc_rx_inplace(&frame, 6, &header);
    if (error) {
        return error;
    }
    *hcho = sensirion_common_bytes_to_int16_t(&buffer[0]);
    *relative_humidity = sensirion_common_bytes_to_int16_t(&buffer[2]);
    *temperature = sensirion_common_bytes_to_int16_t(&buffer[4]);
    return NO_ERROR;
}

int16_t sfa3x_get_device_marking(uint8_t* device_marking,
                                 uint8_t device_marking_size) {
    struct sensirion_shdlc_rx_header header;
    uint8_t buffer[522];
    struct sensirion_shdlc_buffer frame;
    sensirion_shdlc_begin_frame(&frame, &buffer[0], 0xD0, 0x00, 1);
    sensirion_shdlc_add_uint8_t_to_frame(&frame, 0x06);

    sensirion_shdlc_finish_frame(&frame);
    int16_t error = sensirion_shdlc_tx_frame(&frame);
    if (error) {
        return error;
    }

    sensirion_uart_hal_sleep_usec(10000);

    error = sensirion_shdlc_rx_inplace(&frame, 255, &header);
    if (error) {
        return error;
    }
    sensirion_common_copy_bytes(&buffer[0], device_marking,
                                device_marking_size);
    return NO_ERROR;
}

int16_t sfa3x_device_reset() {
    struct sensirion_shdlc_rx_header header;
    uint8_t buffer[12];
    struct sensirion_shdlc_buffer frame;
    sensirion_shdlc_begin_frame(&frame, &buffer[0], 0xD3, 0x00, 0);

    sensirion_shdlc_finish_frame(&frame);
    int16_t error = sensirion_shdlc_tx_frame(&frame);
    if (error) {
        return error;
    }

    sensirion_uart_hal_sleep_usec(200000);

    error = sensirion_shdlc_rx_inplace(&frame, 0, &header);
    if (error) {
        return error;
    }
    sensirion_uart_hal_sleep_usec(200000);
    return NO_ERROR;
}
